### PR TITLE
Ir/compound borrow rate oracle

### PR DESCRIPTION
**Pull request by ioanaroceanu2000**
Compound Borrow Rate Oracle
-- needs some safe math library for the calculations in the `getLastUpdatedRate()` funct

### CODE REVIEW

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        public
+        view
+        override
+        returns (uint32 timestamp, uint256 resultRay)
+    {
+        uint256 borrowIndexPrior = ctoken.borrowIndex();
+
+        uint256 cashPrior = ctoken.getCash();
+        InterestRateModel model = ctoken.interestRateModel();
+        uint256 borrowRateMantissa = model.getBorrowRate(
```
Comment (CyclopsRex) [inline contracts/rate_oracles/CompoundBorrowRateOracle.sol:50]: Is there a reason not to use `CToken.borrowRatePerBlock()`. Would this save us from knowing/caring about the concept of an `InterestRateModel`?

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            ctoken.totalBorrows(),
+            ctoken.totalReserves()
+        );
+        require(
+            borrowRateMantissa <= 0.0005e16,
+            "borrow rate is absurdly high"
+        );
+
+        uint256 blockDelta = block.number - ctoken.accrualBlockNumber();
+        uint256 simpleInterestFactor = borrowRateMantissa * blockDelta;
```
Comment (CyclopsRex) [inline contracts/rate_oracles/CompoundBorrowRateOracle.sol:61]: Maybe we are using simple rather than compound interest because (ironic if true) that's what compound do? Whatever the reason, may be worth an explanatory code comment.

File: deploy/2.mocks.ts [hunk 3]
```diff
@@ -76,6 +76,34 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
...
+
+  const mockCompoundInterestRateModel = await ethers.getContract(
+    "MockCInterestRateModel"
+  );
+
+  const secondsInYear = 31536000;
+  const secondsPerBlock = 13;
+  const ratePerBlock = BigNumber.from(10)
+    .pow(16)
+    .mul(0.02 * 100)
```
Comment (CyclopsRex) [inline deploy/2.mocks.ts:96]: Worth defining constants (or adding explanatory comment) for these magic values

File: tasks/getHistoricalData.ts [hunk 4]
```diff
@@ -91,6 +91,12 @@ task("getHistoricalData", "Retrieves the historical rates")
     "ETH",
     types.string
   )
+  .addOptionalParam(
```
Comment (CyclopsRex) [inline tasks/getHistoricalData.ts:94]: Was this copied from Sol's code? May now be out of date cos I think he was going to switch to a flag.

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    const testRateOracle = (await testRateOracleFactory.deploy(
+      cToken.address,
+      false,
+      token.address,
+      await token.decimals()
+    )) as TestRateOracle;
+    return { testRateOracle };
+  },
+  setRateAsDecimal: async (rate: number) => {
+    // To set the rate for compound, we call setExchangeRate on the cToken
```
Comment (CyclopsRex) [inline test/main_contracts/rate_oracles/compoundBorrowConfig.ts:50]: Comment out of date

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      await mockCInterestRateModel.setBorrowRatePerBlock(ratePerBlock);
+    });
+    it("Verify correct protocol ID for Compound Borrow rate oracle", async () => {
+      const protocolID =
+        await testCompoundBorrowRateOracle.UNDERLYING_YIELD_BEARING_PROTOCOL_ID();
+
+      expect(protocolID).to.eq(2);
+    });
+
+    it(`Verify rate for Compound Borrow`, async () => {
```
Comment (CyclopsRex) [inline test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:48]: The logic for `getLastUpdatedRate` is more involved than other implementations and if it stays that way there is probably more we should test here. E.g. testing with a variety of different blockDelta values, including 0 and 1 as well as large numbers.

**Review by CyclopsRex — COMMENTED** [review]
Added some comments.

File: contracts/interfaces/compound/InterestRateModel.sol [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity =0.8.9;
+
+/**
+  * @title Compound's InterestRateModel Interface
+  * @author Compound
+  */
+abstract contract InterestRateModel {
```
[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/interfaces/compound/InterestRateModel.sol:8]: agreed, @ioanaroceanu2000 not sure if we need isInterestRateModel either

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import "../interfaces/compound/InterestRateModel.sol";
+
+contract CompoundBorrowRateOracle is BaseRateOracle, ICompoundRateOracle {
+    /// @inheritdoc ICompoundRateOracle
+    ICToken public immutable override ctoken;
+
+    /// @inheritdoc ICompoundRateOracle
+    uint8 public immutable override decimals;
+
+    uint8 public constant override UNDERLYING_YIELD_BEARING_PROTOCOL_ID = 2; // id of compound is 2
```
Comment (arturbeg) [inline contracts/rate_oracles/CompoundBorrowRateOracle.sol:17]: the id should be 6 for the compound borrow rate oracle since 2 is reserved for the compound supply rate oracle

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 1 above)
Comment (arturbeg) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:50]: @ioanaroceanu2000 i think what @CyclopsRex is suggesting makes sense, it seems like https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L207 

is doing the same thing anyway, feels like we can safely get rid of the interest rate model

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+        uint256 cashPrior = ctoken.getCash();
+        InterestRateModel model = ctoken.interestRateModel();
+        uint256 borrowRateMantissa = model.getBorrowRate(
+            cashPrior,
+            ctoken.totalBorrows(),
+            ctoken.totalReserves()
+        );
+        require(
+            borrowRateMantissa <= 0.0005e16,
```
[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:56]: @CyclopsRex seems like its coming from here 

https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L345

[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:56]: @ioanaroceanu2000 i think if we are using a constant such as 0.0005e16 it is better to define it as a constant and then reuse the constant name for readability similar to how it is done in compound smart contracts, wdyt @CyclopsRex ?

https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CTokenInterfaces.sol#L31

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 2 above)
Comment (arturbeg) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:61]: yeah, compound is using a simple interest rate factor:) worth adding comments though, and perhaps adding a reference to the permalink below:

https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L359

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            "borrow rate is absurdly high"
+        );
+
+        uint256 blockDelta = block.number - ctoken.accrualBlockNumber();
+        uint256 simpleInterestFactor = borrowRateMantissa * blockDelta;
+
+        uint256 expScale = 1e18;
+        uint256 borrowIndex = (simpleInterestFactor * borrowIndexPrior) /
```
[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:64]: Re:

https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L363

@ioanaroceanu2000 i think it is worth using the same maths methods as compound for this calculation by importing this https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/ExponentialNoError.sol#L45, wdyt @CyclopsRex ?

File: deploy/2.mocks.ts (hunk 3 above)
Comment (arturbeg) [reply deploy/2.mocks.ts:96]: agreed, @ioanaroceanu2000 just to confirm this is only relevant for testing purposes of the mocks?

File: tasks/getHistoricalData.ts (hunk 4 above)
Comment (arturbeg) [reply tasks/getHistoricalData.ts:94]: @ioanaroceanu2000 i think @Sol3iman mentioned that you had a catchup with him on how best to align your code with his, worth rebasing his changes on top of yours as Sol's code is now merged into main and make sure there is alignment

File: tasks/getHistoricalData.ts [hunk 11]
```diff
@@ -289,19 +295,32 @@ task("getHistoricalData", "Retrieves the historical rates")
...
             if (cToken && decimals) {
-              let r = await cToken.exchangeRateStored({
-                blockTag: b,
-              });
-              if (decimals > 17) {
-                r = r.div(BigNumber.from(10).pow(decimals - 17));
-              } else if (decimals < 17) {
-                r = r.mul(BigNumber.from(10).pow(17 - decimals));
+              if (taskArgs.rate === "borrow") {
+                let r = await cToken.borrowIndex({
```
Comment (arturbeg) [inline tasks/getHistoricalData.ts:299]: @ioanaroceanu2000 does this logic take care of interpolation? in other words is it one to one with the compound borrowing rate oracle logic we have in the smart contracts?

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    // store the cToken and interest rate model and some rate per block info for use when setting rates
+    cToken = (await ethers.getContract("MockCToken")) as MockCToken;
+    const mockCInterestRateModel = (await ethers.getContract(
+      "MockCInterestRateModel"
+    )) as MockCInterestRateModel;
+
+    // Set rate per block in IR model -> 2% per year
+    // const blocksPerYear = toBn(31536000).div(toBn(13));
+    // const ratePerBlock = toBn(0.02, 18).div(blocksPerYear);
+    await mockCInterestRateModel.setBorrowRatePerBlock(0);
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowConfig.ts:29]: do we want to set this to 0, doesn't seem to match the comments above?

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        "MockCInterestRateModel"
+      )) as MockCInterestRateModel;
+
+      await mockCInterestRateModel.setBorrowRatePerBlock(ratePerBlock);
+    });
+    it("Verify correct protocol ID for Compound Borrow rate oracle", async () => {
+      const protocolID =
+        await testCompoundBorrowRateOracle.UNDERLYING_YIELD_BEARING_PROTOCOL_ID();
+
+      expect(protocolID).to.eq(2);
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:45]: we should expect the protocol id to be 6

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 6 above)
Comment (arturbeg) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:48]: agreed, @CyclopsRex can you pls write up a set of edge case scenarios you'd personally test in this instance? the main difference between compound and aave borrowing rate oracles is we manually have to do borrow index extrapolation using the current borrow rate + block delta between the latest accrual on compound and current block

**Review by arturbeg — CHANGES_REQUESTED** [review]

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 1 above)
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:50]: I missed that function
Makes a lot of sense, I'll replace it

File: contracts/interfaces/compound/ICToken.sol [hunk 14]
```diff
@@ -2,28 +2,66 @@
 
 pragma solidity =0.8.9;
 
+import "./InterestRateModel.sol";
+
 // Subset of https://github.com/compound-finance/compound-protocol/blob/master/contracts/CTokenInterfaces.sol
-interface ICToken {
+contract CTokenStorage {
```
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/interfaces/compound/ICToken.sol:8]: Because we use storage now we can't have this as an interface. The Borrow Oracle calls getter functions (e.g. https://github.com/Voltz-Protocol/voltz-core/blob/ir/compound-borrow-rate-oracle/contracts/rate_oracles/CompoundBorrowRateOracle.sol#L46). We can rename this to "CTokenInterface" as Compound did to avoid confusion with an actual interface... what do you think?

File: contracts/interfaces/compound/InterestRateModel.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/interfaces/compound/InterestRateModel.sol:8]: due to the suggestion of using ICToken's getBorrowatePerBlock(), I removed this entire contract

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:56]: Yes, using `borrowRateMaxMantissa`

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        );
+        require(
+            borrowRateMantissa <= 0.0005e16,
+            "borrow rate is absurdly high"
+        );
+
+        uint256 blockDelta = block.number - ctoken.accrualBlockNumber();
+        uint256 simpleInterestFactor = borrowRateMantissa * blockDelta;
+
+        uint256 expScale = 1e18;
```
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:63]: removed

File: deploy/2.mocks.ts (hunk 3 above)
Comment (ioanaroceanu2000) [reply deploy/2.mocks.ts:110]: yes, only relevant for testing. This is not set manually in real scenarios

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 2 above)
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:54]: agreed

File: contracts/interfaces/compound/ICToken.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/interfaces/compound/ICToken.sol:6]: I removed the storage contract and added getters in the `ICToken` definition. We can make it an interface now to follow best practices.  The `borrowRateMaxMantissa` is now defined in oracle itself to be used in the require statement `require(borrowRateMantissa <= borrowRateMaxMantissa,"borrow rate is absurdly high");`

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+            Exp({mantissa: borrowRateMantissa}),
+            blockDelta
+        );
+
+        uint256 borrowIndexPrior = ctoken.borrowIndex();
+        uint256 borrowIndex = mul_ScalarTruncateAddUInt(
+            simpleInterestFactor,
+            borrowIndexPrior,
+            borrowIndexPrior
+        ); // result given in wad, scale to ray
```
Comment (arturbeg) [inline contracts/rate_oracles/CompoundBorrowRateOracle.sol:74]: can we re-confirm that the result is indeed given in wad?

e.g. looking through the supply compound rate oracle, it appears the conversion is slightly more involved as it is a function of the number of decimals of the underlying token:

https://github.com/Voltz-Protocol/voltz-core/blob/ac45489df29306b2e648dccb940d6c6eb6c2a406/contracts/rate_oracles/CompoundRateOracle.sol#L75

wdyt @CyclopsRex ?

File: contracts/utils/ExponentialNoError.sol [hunk 17]
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Comment (arturbeg) [inline contracts/utils/ExponentialNoError.sol:1]: https://en.wikipedia.org/wiki/BSD_licenses

"The BSD license is a simple license that merely requires that all code retain the BSD license notice if redistributed in source code format, or reproduce the notice if redistributed in binary format. The BSD license (unlike some other licenses e.g. [GPL](https://en.wikipedia.org/wiki/GPL)) does not require that source code be distributed at all.

seems okay to me, thoughts @CyclopsRex ?

File: deploy/2.mocks.ts [hunk 18]
```diff
@@ -90,6 +90,17 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
     await trx.wait();
   }
 
+  // set initial borrow index
+  let trx = await mockCToken.setBorrowIndex(BigNumber.from(10).pow(18));
+  await trx.wait();
+
+  const blocksPerYear = BigNumber.from(31536000).div(13);
+  const wad = BigNumber.from(10).pow(18);
```
Comment (arturbeg) [inline deploy/2.mocks.ts:98]: @ioanaroceanu2000 can you pls confirm that these numbers are indeed supposed to be in wad and are not a function of the number of decimals of the underlying token?

File: tasks/getHistoricalData.ts [hunk 19]
```diff
@@ -288,28 +286,52 @@ task("getHistoricalData", "Retrieves the historical rates")
...
             if (cToken && decimals) {
-              let r = await cToken.exchangeRateStored({
-                blockTag: b,
-              });
-              if (decimals > 17) {
-                r = r.div(BigNumber.from(10).pow(decimals - 17));
-              } else if (decimals < 17) {
-                r = r.mul(BigNumber.from(10).pow(17 - decimals));
-              }
+              if (taskArgs.borrow) {
```
Comment (arturbeg) [inline tasks/getHistoricalData.ts:296]: @ioanaroceanu2000 will be helpful to see the apy and rate value plots generated by this script and compare them against compound ui and loanscan historical borrow apy data

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  oracleFixture: async () => {
+    // Use hardhat-deploy to deploy factory and mocks
+    await deployments.fixture(["Factory", "Mocks"]);
+    const token = await ethers.getContract("ERC20Mock");
+
+    // store the cToken and interest rate model and some rate per block info for use when setting rates
+    cToken = (await ethers.getContract("MockCToken")) as MockCToken;
+
+    // Set inital rate per block to 0
+    await cToken.setBorrowRatePerBlock(0);
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowConfig.ts:20]: do we want to set this to a more realistic value?

Comment (arturbeg) [reply test/main_contracts/rate_oracles/compoundBorrowConfig.ts:20]: to make sure the interpolation works as expected

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+    // set initial borrow index
+    await cToken.setBorrowIndex(toBn(startingExchangeRate, 18));
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowConfig.ts:23]: again, will be helpful to confirm that the borrow index is indeed in wad terms and is not a function of the number of decimals in the underlying token

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+    // set last update block
+    const latestBlock = await ethers.provider.getBlock("latest");
+    await cToken.setAccrualBlockNumber(latestBlock.number);
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowConfig.ts:27]: this setup means there is no interpolation, are we able to test the interpolation logic works as expected?

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+import { ConfigForGenericTests as Config } from "./compoundBorrowConfig";
+import { MockCToken, TestCompoundBorrowRateOracle } from "../../../typechain";
+import { advanceTimeAndBlock } from "../../helpers/time";
+
+const { provider } = waffle;
+
+describe("Compound Borrow Rate Oracle", () => {
+  const blocksPerYear = BigNumber.from(31536000).div(13);
+  const wad = BigNumber.from(10).pow(18);
+  const halfWad = BigNumber.from(10).pow(9);
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:13]: hmm,  is halfWad not equal to BigNumber.from(5).pow(17)? wdyt @0xZenus ?

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts [hunk 24]
```diff
@@ -0,0 +1,129 @@
+import { ethers, waffle } from "hardhat";
```
Comment (arturbeg) [inline test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:1]: @CyclopsRex @0xZenus this needs a very careful review

**Review by arturbeg — CHANGES_REQUESTED** [review]

File: contracts/utils/ExponentialNoError.sol (hunk 17 above)
Comment (CyclopsRex) [reply contracts/utils/ExponentialNoError.sol:1]: License seems OK. I want to take a closer look at how this is being used, given we already have two libraries for working with decimals represented as wad so it would be nice to avoid a third.

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 23 above)
Comment (CyclopsRex) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:13]: I think it's neither of these things, but is `BigNumber.from(10).pow(17).mul(5)` 😄

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (CyclopsRex) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol]: Yes, sorry, I noticed that's where it came from and forgot to delete this comment.

Do I understand correctly that we think that compound can never return a value in excess of 0.0005e16, so this is just a sanity check and not actually introducing any tighter restrictions than already exist? If so, please confirm that you think it's a worthwhile sanity check.

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 6 above)
Comment (CyclopsRex) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts]: The thing that's new/different is this extrapolation by blocks-since-last-accrual, so I would test that for a variety of `blockDelta` values, including `blockDelta = 0` and `blockDelta = 1`. Check that the math works out as we expect.

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (CyclopsRex) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol]: Thanks. It's subjective cos there's certainly value in having code that's as close as possible to Compund's code, but I personally would favour using our existing 10^18 maths libraries rather than introducing a third one to do basically the same thing.

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 16 above)
Comment (CyclopsRex) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:74]: Yeah, I too am a little surprised if this is in wad but the maths Compound is doing does seem to imply that. Worth checking for sure.

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 24 above)
Comment (CyclopsRex) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:1]: It seems OK but:

1. it might be worth moving the logic to calculate the expected rate into its own function for easy re-use
 
2. I would feel more comfortable if we were also doing some testing from first principles. E.g.

 - choose an APY, `a`
 - set the block rate as we think is required to effect that APY
 - write some rates with various block deltas
 - check that the APY we calculate pairwise between any of the chosen written data points is equal to `a`, perhaps with some small allowance for compounding.

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 23 above)
Comment (ioanaroceanu2000) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:13]: naming problem... it's supposed to be 1e9. Is `semiWad` more representative of this?

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol]: we followed the implementation from compound's `accrueInterest` function and whenever they acrue interest, they don't allow a "ratePerBlock" higher than this value. To stay accurate, we should include this validation

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 6 above)
Comment (ioanaroceanu2000) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts]: test cases included

File: tasks/getHistoricalData.ts (hunk 11 above)
Comment (ioanaroceanu2000) [reply tasks/getHistoricalData.ts]: it was somehow lost in the rebasing process. I replaced it by the interpolation logic now

File: contracts/rate_oracles/CompoundBorrowRateOracle.sol (hunk 16 above)
Comment (ioanaroceanu2000) [reply contracts/rate_oracles/CompoundBorrowRateOracle.sol:74]: Screenshot 2022-08-03 at 19 37 42
taken from cDAI
Screenshot 2022-08-03 at 19 38 50
taken from cUSDC

File: deploy/2.mocks.ts (hunk 18 above)
Comment (ioanaroceanu2000) [reply deploy/2.mocks.ts:98]: https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L205

File: tasks/getHistoricalData.ts (hunk 19 above)
Comment (ioanaroceanu2000) [reply tasks/getHistoricalData.ts:296]: you can find those comparisons here https://www.notion.so/voltz/Borrowing-Compound-Rate-Oracle-Sanity-Checks-b35fe7889309494a9c453f083cba6002

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts (hunk 20 above)
Comment (ioanaroceanu2000) [reply test/main_contracts/rate_oracles/compoundBorrowConfig.ts:20]: the generic tests assume that the initial rate is 1 (which is set through the `borrowIndex`). tho we check if the interpolation works in the compound borrow tests - https://github.com/Voltz-Protocol/voltz-core/blob/ir/compound-borrow-rate-oracle/test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 24 above)
Comment (ioanaroceanu2000) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts:1]: I implemented this test using a series of different block deltas and rates

File: test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts (hunk 23 above)
Comment (arturbeg) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts]: oh yeah, ofc lol

Comment (arturbeg) [reply test/main_contracts/rate_oracles/compoundBorrowRateOracle.ts]: oh i see, yeah semiWad is fine:)

File: test/main_contracts/rate_oracles/compoundBorrowConfig.ts (hunk 20 above)
Comment (arturbeg) [reply test/main_contracts/rate_oracles/compoundBorrowConfig.ts:20]: got it
